src/topology/metric_space/glueing.lean

```diff
@@ -0,0 +1,357 @@
+      exists_lt_of_cInf_lt (by simp [_inst_1]) (by rw [infi]; linarith),
+    rcases this with ⟨a, arange, ha⟩,
+    rcases mem_range.1 arange with ⟨p, pa⟩,
+    rw ← pa at ha,
+    have : ∃b ∈ range (λp, dist z (Φ p) + dist y (Ψ p)), b < infi (λp, dist z (Φ p) + dist y (Ψ p)) + δ/2 :=
+      exists_lt_of_cInf_lt (by simp [_inst_1]) (by rw [infi]; linarith),
+    rcases this with ⟨b, brange, hb⟩,
+    rcases mem_range.1 brange with ⟨q, qb⟩,
+    rw ← qb at hb,
+    have : dist (Φ p) (Φ q) ≤ dist (Ψ p) (Ψ q) + 2 * ε,
+      { have := le_trans (le_abs_self _) (H p q), by linarith },
+    calc dist x z ≤ dist x (Φ p) + dist (Φ p) (Φ q) + dist (Φ q) z : dist_triangle4 _ _ _ _
+      ... ≤ dist x (Φ p) + dist (Ψ p) (Ψ q) + dist z (Φ q) + 2 * ε : by rw [dist_comm z]; linarith
+      ... ≤ dist x (Φ p) + (dist y (Ψ p) + dist y (Ψ q)) + dist z (Φ q) + 2 * ε :
+        add_le_add (add_le_add (add_le_add (le_refl _) (dist_triangle_left _ _ _)) (le_refl _)) (le_refl _)
+      ... ≤ (infi (λp, dist x (Φ p) + dist y (Ψ p)) + ε) + (infi (λp, dist z (Φ p) + dist y (Ψ p)) + ε) + δ :
+        by linarith
+  end
+| (inr x) (inl y) (inr z) := real.le_of_forall_epsilon_le $ λδ δpos, begin
+    have : ∃a ∈ range (λp, dist y (Φ p) + dist x (Ψ p)), a < infi (λp, dist y (Φ p) + dist x (Ψ p)) + δ/2 :=
+      exists_lt_of_cInf_lt (by simp [_inst_1]) (by rw [infi]; linarith),
+    rcases this with ⟨a, arange, ha⟩,
+    rcases mem_range.1 arange with ⟨p, pa⟩,
+    rw ← pa at ha,
+    have : ∃b ∈ range (λp, dist y (Φ p) + dist z (Ψ p)), b < infi (λp, dist y (Φ p) + dist z (Ψ p)) + δ/2 :=
+      exists_lt_of_cInf_lt (by simp [_inst_1]) (by rw [infi]; linarith),
+    rcases this with ⟨b, brange, hb⟩,
+    rcases mem_range.1 brange with ⟨q, qb⟩,
+    rw ← qb at hb,
+    have : dist (Ψ p) (Ψ q) ≤ dist (Φ p) (Φ q) + 2 * ε,
+      { have := le_trans (neg_le_abs_self _) (H p q), by linarith },
+    calc dist x z ≤ dist x (Ψ p) + dist (Ψ p) (Ψ q) + dist (Ψ q) z : dist_triangle4 _ _ _ _
+      ... ≤ dist x (Ψ p) + dist (Φ p) (Φ q) + dist z (Ψ q) + 2 * ε : by rw [dist_comm z]; linarith
+      ... ≤ dist x (Ψ p) + (dist y (Φ p) + dist y (Φ q)) + dist z (Ψ q) + 2 * ε :
+        add_le_add (add_le_add (add_le_add (le_refl _) (dist_triangle_left _ _ _)) (le_refl _)) (le_refl _)
+      ... ≤ (infi (λp, dist y (Φ p) + dist x (Ψ p)) + ε) + (infi (λp, dist y (Φ p) + dist z (Ψ p)) + ε) + δ :
+        by linarith
+  end
+
+private lemma glue_eq_of_dist_eq_zero (Φ : γ → α) (Ψ : γ → β) (ε : ℝ) (ε0 : 0 < ε) :
+  ∀p q : α ⊕ β, glue_dist Φ Ψ ε p q = 0 → p = q
+| (inl x) (inl y) h := by rw eq_of_dist_eq_zero h
+| (inl x) (inr y) h := begin
+    have : 0 ≤ infi (λp, dist x (Φ p) + dist y (Ψ p)) :=
+      le_cinfi (λp, by simpa using add_le_add (@dist_nonneg _ _ x _) (@dist_nonneg _ _ y _)),
+    have : 0 + ε ≤ glue_dist Φ Ψ ε (inl x) (inr y) := add_le_add this (le_refl ε),
+    have : false, by linarith,
```

Comment:
Maybe `exfalso, linarith` to finish the proof would be a bit more idiomatic.

Reply:
Done